### PR TITLE
Adding polyline and polygon methods

### PR DESCRIPTION
**Pull request by uovodikiwi**
As per issue [#58](https://github.com/PyFPDF/fpdf2/issues/58). 
I will probably need to add the documentation of this methods.

### CODE REVIEW

**Comment by uovodikiwi** [comment]
@Lucas-C I will probably need to change something :) let me know

**Comment by alexp1917** [comment]
needs tests :) i can try to make those tonight but we dont have write access to the same repositories so ill just be able to post a test file in a comment or something.

**Comment by uovodikiwi** [comment]
@alexp1917 No worries! I will add them in the following days. Just wanted to be sure that the code-style was ok :) Thank you for your comment

**Comment by alexp1917** [comment]
@uovodikiwi I got a little bit lazy towards the end with the "end-to-end" actual pdf-producing tests, but here's something to start.

As i was doing this, it occurred to me that these methods should definitely eventually take a thickness and a color, (that doesn't have to happen in this PR).

<details>
<summary>

`test/shapes/test_polyline.py`

</summary>

```python
from pathlib import Path

import fpdf
from fpdf.errors import FPDFException
from test.utilities import assert_pdf_equal

import pytest

HERE = Path(__file__).resolve().parent


def with_page_and_data(unit):
    pdf = fpdf.FPDF(unit=unit)
    pdf.add_page()
    data = []
    # pylint: disable=protected-access
    pdf._out = data.append
    return pdf, data


polyline_coordinates = [(10, 10), (40, 10), (10, 40)]
expected_pdf_str = "10.00 831.89 m40.00 831.89 l10.00 801.89 l S "
expected_pdf_str_fill = "10.00 831.89 m40.00 831.89 l10.00 801.89 l B "
expected_pdf_str_poly = "10.00 831.89 m40.00 831.89 l10.00 801.89 l h  S "
expected_pdf_str_polyfill = "10.00 831.89 m40.00 831.89 l10.00 801.89 l h  B "


def scale_points(raw_points, k_recip):
    return [(k_recip * coord[0], k_recip * coord[1]) for coord in raw_points]


scaling_factors_for_units = [
    ("pt", 1),
    ("mm", 1 / (72 / 25.4)),
    ("cm", 1 / (72 / 2.54)),
    ("in", 1 / 72),
]


@pytest.mark.parametrize("k,factor", scaling_factors_for_units)
def test_polyline_command_all_k(k, factor):
    pdf, data = with_page_and_data(k)
    pdf.polyline(scale_points(polyline_coordinates, factor))
    assert expected_pdf_str == "".join(data)

    data.clear()

    pdf.polyline(scale_points(polyline_coordinates, factor), fill=True)
    assert expected_pdf_str_fill == "".join(data)

    data.clear()

    pdf.polyline(scale_points(polyline_coordinates, factor), polygon=True)
    assert expected_pdf_str_poly == "".join(data)

    data.clear()
    pdf.polyline(scale_points(polyline_coordinates, factor), polygon=True, fill=True)
    assert expected_pdf_str_polyfill == "".join(data)


def test_check_page():
    pdf = fpdf.FPDF(unit="pt")
    with pytest.raises(FPDFException) as polyline_no_page:
        pdf.polyline(polyline_coordinates)

    msg = "No page open, you need to call add_page() first"
    assert msg == str(polyline_no_page.value)

    with pytest.raises(FPDFException) as polygon_no_page:
        pdf.polygon(polyline_coordinates)

    assert msg == str(polygon_no_page.value)


def test_polyline_fill_polygon(tmp_path):
    pdf = fpdf.FPDF(unit="cm")
    pdf.add_page()
    pdf.set_x(10)
    pdf.set_y(10)
    coords = scale_points(polyline_coordinates, scaling_factors_for_units[2][1])
    pdf.polyline(coords, fill=True, polygon=True)
    # pdf.output(HERE / "class_polyline.pdf")
    # uncomment me to produce the file, then commit result into the repository so the next line succeeeds
    assert_pdf_equal(pdf, HERE / "class_polyline.pdf", tmp_path)
```
</details>

**Comment by alexp1917** [comment]
its also on the `polyline-method_tmp` branch on here for convenience

File: fpdf/fpdf.py [hunk 1]
```diff
@@ -52,7 +52,7 @@
 HERE = Path(__file__).resolve().parent
 
 # Global variables
-FPDF_VERSION = "2.3.0"
+FPDF_VERSION = "2.3.1"
```
Comment (Lucas-C) [inline fpdf/fpdf.py:55]: Please do not bump the lib version number as part of a PR.
Maintainers will increment it on the `master` branch when a new release will be performed.

**Review by Lucas-C — CHANGES_REQUESTED** [review]
Good job @uovodikiwi, thank you very much for your contribution!
The code & docstrings are perfect.

I agree with @alexp1917 about the need for some tests.
Based on those he suggested, I created a new branch, rebased from `master`, with documentation & tests: https://github.com/PyFPDF/fpdf2/commits/pull-105

@uovodikiwi: would you prefer to finish the work on this PR yourself, or should we take over so that this great addition can be merged in the next release? 😊

**Comment by Lucas-C** [comment]
I slightly modified your last test @alexp1917 in order to draw a pentagon instead of a triangle
(just because I really wanted to try drawing a pentagon ^^):
https://github.com/PyFPDF/fpdf2/blob/5bd99d8eee402ff890a9fa0891edbbc77ffb7437/test/shapes/filled_polygon.pdf

**Comment by alexp1917** [comment]
@Lucas-C @uovodikiwi it seems like if the pull-105 branch is squashed and merged, then GitHub will assign all committers equal credit so it seems it would still be including the contribution (at minimum it would be two committers on the merge b/c a maintainer needs to merge into main). this seems fine to me and if @uovodikiwi you dont mind we'll give it a couple days to hear back or else just merge the pull-105 branch.

@Lucas-C it was intended that my test be modified at the very least to include all the branches in the new code.

this seems fairly ready, and then afterwards it is worth revisiting either the svg roadmap or a new release?

**Comment by uovodikiwi** [comment]
> @Lucas-C @uovodikiwi it seems like if the pull-105 branch is squashed and merged, then GitHub will assign all committers equal credit so it seems it would still be including the contribution (at minimum it would be two committers on the merge b/c a maintainer needs to merge into main). this seems fine to me and if @uovodikiwi you dont mind we'll give it a couple days to hear back or else just merge the pull-105 branch.
> 
> @Lucas-C it was intended that my test be modified at the very least to include all the branches in the new code.
> 
> this seems fairly ready, and then afterwards it is worth revisiting either the svg roadmap or a new release?

Sorry guys :( I had a couple of rough weeks and couldn't work on personal projects. For me it's ok, let me know if I can something more for this!

**Comment by alexp1917** [comment]
@uovodikiwi thanks for the quick reply, yeah thats kind of how it is with open source, at various times people help out when they can and its appreciated and merged. we'll get this into the main branch and stay tuned for more issues and stuff i suppose.

**Comment by alexp1917** [comment]
closing in favor of #113

**Pull request closed without merging**